### PR TITLE
Add .env file parsing, fix laravel to use it, fix Shopware 6 overwriting, fixes #3636, fixes #3590

### LINKED ISSUE #3590: Project type shopware6 keeps overwriting APP_URL
**Run a Diagnostic and Paste Link Here**
https://gist.github.com/susannemoog/2b36c2feb9746392606c6eedf4353d0c

**Describe the bug**
With a project of type shopware6 manual changes to the APP_URL are not possible, as every restart adds an additional line with the original ddev created value.

**To Reproduce**
```
name: shopware
type: shopware6
docroot: public
php_version: "7.4"
webserver_type: nginx-fpm
router_http_port: "80"
router_https_port: "443"
xdebug_enabled: false
additional_hostnames: []
additional_fqdns: []
mariadb_version: "10.3"
mysql_version: ""
nfs_mount_enabled: false
mutagen_enabled: false
use_dns_when_possible: true
composer_version: "2"
web_environment: []
```

- Add APP_URL manually to .env
- (Re)Start ddev 
- Open .env
- Adjust APP_URL ddev created to http
- Restart ddev

Results in something like this in the .env:
```
APP_ENV="dev"
APP_URL="http://shopware.ddev.site"
...
# =================
# Configuration added by ddev
APP_URL="http://shopware.ddev.site"
# =================
# Configuration added by ddev
APP_URL="http://shopware.ddev.site"
# =================
# Configuration added by ddev
APP_URL="http://shopware.ddev.site"
# =================
# Configuration added by ddev
APP_URL="https://shopware.ddev.site"
```

**Expected behavior**

- If APP_URL is already set, do not add another one

**Version and configuration information (please complete the following information):**

config.yaml - see above
version info - see gist above
Shopware installation according to https://susi.dev/ddev-shopware-6/

### LINKED ISSUE #3636: Improve Laravel's explicit integration by actually updating the .env file
**Is your feature request related to a problem? Please describe.**

DDEV does a good job of setting up itself and pre-configuring a Drupal project.
It would be nice if we could automate other frameworks too. 

**Describe the solution you'd like**
After running `ddev config` in a new Laravel project, there are 2 additional steps required.
- Updating PHP version. Addressed in #3588 
- Updaitng Laravel `.env` file

```diff
# These are the current defaults
APP_URL=http://localhost
DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=laravel
DB_USERNAME=root
DB_PASSWORD=
```

Eg. Assuming I create  a new project call phpcs-demo 

## Test

- Start a new `Laravel` project

```shell
laravel new phpcs-demo
cd phpcs-demo
```

- Add `DDEV` to project, accepting all default

```shell
$ ddev config
...
Configuration complete. You may now run 'ddev start'.
```

- Bump DDEV `PHP` version in `./.ddev/config.yaml`

```yaml
php_version: "8.0"
```

- Config Laravel `.env`

```env
APP_URL=https://phpcs-demo.ddev.site
DB_CONNECTION=mysql
DB_HOST=ddev-phpcs-demo-db
DB_DATABASE=db
DB_USERNAME=db
DB_PASSWORD=db
```

This assumes a `mysql` database, but I _think_ all of these might be ddev variables already?
It would get great if a laravel-project type ddev config could automatically update the `.env` file too.

**Additional context**
Ideally, we would want to update other supported frameworks (WP, Magento) but I have no idea how involved those are. A Laravel feels like an easy win to start though.

### PR DESCRIPTION
**Pull request by rfay**
## The Problem/Issue/Bug:

We need to read, write, and support .env files better for various project types.

* #3636
* #3590

## How this PR Solves The Problem:

* Add reading and writing of .env
* For Laravel, directly edit the .env file
* For Shopware 6, mostly stop editing it except MAILER_URL

## Manual Testing Instructions:

- [x] Do manual installation of Laravel and then use it
- [x] Do manual installation of Shopware 6 and then use it
- [x] Verify that only appropriate changes are made to the .env

## Automated Testing Overview:

No changes, existing should catch it.

<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/4197"></a>

### CODE REVIEW

File: pkg/ddevapp/laravel.go [hunk 1]
```diff
@@ -18,35 +19,31 @@ func isLaravelApp(app *DdevApp) bool {
...
+		}
+		envContents, err = ReadEnvFile(app)
+		if err != nil {
+			return err
 		}
-	} else {
-		envSettingsWarning(WarnTypeAbsent)
+	}
+	envContents["DB_HOST"] = "db"
+	envContents["DB_PORT"] = "3306"
```
Comment (tyler36) [inline pkg/ddevapp/laravel.go:38]: This assumes a "MySql" database type. Doesn't DDEV have a DB_PORT variable we could use instead?

File: pkg/ddevapp/laravel.go [hunk 2]
```diff
@@ -18,35 +19,31 @@ func isLaravelApp(app *DdevApp) bool {
...
+	envContents["DB_DATABASE"] = "db"
+	envContents["DB_USERNAME"] = "db"
+	envContents["DB_PASSWORD"] = "db"
+	envContents["DB_CONNECTION"] = "ddev"
```
Comment (tyler36) [inline pkg/ddevapp/laravel.go:42]: Setting the "DB_CONNECTION" here is invalid for Laravel unless clients add a DDEV section to `./config/database.php`. That requires a big chunk of code. 

The solution I came up with for the action PR (https://github.com/drud/ddev/pull/3954/files) is more elegant I think; use the database name. This is more future proof too, if Laravel updates the database options (strict, charset etc.).

Note that "postgres" uses a "pgsql" in the laravel config file.

File: pkg/ddevapp/laravel.go (hunk 1 above)
Comment (rfay) [reply pkg/ddevapp/laravel.go:38]: I did some quick look-around and didn't see support for postgres on Laravel. Does it have it?

File: pkg/ddevapp/laravel.go (hunk 2 above)
Comment (rfay) [reply pkg/ddevapp/laravel.go:42]: So great that you have studied this so well! I think we'll have to talk in Discord for me to understand the big picture.

File: pkg/ddevapp/laravel.go (hunk 1 above)
Comment (tyler36) [reply pkg/ddevapp/laravel.go:38]: It definately has it: https://laravel.com/docs/9.x/database#introduction
My company tends to use Postgres for most of its projects, it was one of the reason I pushed for the ddev-contrib module to get first-party support in DDEV.

See https://github.com/drud/ddev/pull/4197#discussion_r966558313

**Comment by rfay** [comment]
@susannemoog Would love it if you could take a look and test this. Artifacts at https://github.com/drud/ddev/pull/4197#issuecomment-1241409569

**Comment by rfay** [comment]
Sadly, I think I'm going to have to give up on this for now. It has too many flaws and too much water under the bridge. It shouldn't be terribly hard to redo using @khalwat 's settings regex replace technique from Craft CMS implementation, and referring to @tyler36 's notes here.

If anybody would be interested in pushing this through I'd sure be happy to help.

**Comment by rfay** [comment]
Giving this another try.

File: docs/content/users/quickstart.md [hunk 3]
```diff
@@ -321,8 +321,6 @@ DDEV comes ready to work with any PHP project, and has deeper support for severa
         ddev config --project-type=laravel --docroot=public --create-docroot
```
Comment (tyler36) [inline docs/content/users/quickstart.md:321]: This workflow results in an error:

```shell
ddev start
Starting my-laravel-app...
Network ddev-my-laravel-app_default  Created
Container ddev-my-laravel-app-dba  Started
Container ddev-my-laravel-app-web  Started
Container ddev-my-laravel-app-db  Started
Container ddev-router  Running
Failed to start my-laravel-app: open /home/user13/code/playground/my-laravel-app/.env.example: no such file or directory
```

`.env.example` is created by the `ddev composer create --prefer-dist laravel/laravel`

Comment (rfay) [reply docs/content/users/quickstart.md:321]: Ah, and our `ddev start` is happening earlier, thanks. I must have tested with an existing setup.

Comment (tyler36) [reply docs/content/users/quickstart.md:321]: The clone instructions are fine.

**Comment by tyler36** [comment]
- ✅  Test Laravel clone with Postgres database

```shell
ddev config --project-type=laravel --docroot=public --create-docroot --database=postgres:14
```

**Comment by tyler36** [comment]
Is there a reason we don't update `APP_URL` for Laravel?

The original instructions included setting `APP_URL` to `$DDEV_PRIMARY_URL` in `.env`.

**Comment by rfay** [comment]
Note: Manually testing Laravel resulted in a failure which goes back to 
* https://github.com/drud/ddev/pull/4273#issuecomment-1335649935

As a result, I added a [commit](https://github.com/drud/ddev/pull/4197/commits/12143e41b08465cd004f58b66d87c9e58ad064d4) to update the Laravel quickstart (more) to do `ddev composer create --no-scripts` followed by `ddev composer install`

File: docs/content/users/quickstart.md (hunk 3 above)
Comment (rfay) [reply docs/content/users/quickstart.md]: I made it accept if the .env.example isn't there and just skip the whole thing. Great catch.

**Comment by rfay** [comment]
Thanks for all the feedback. I think it's all covered now. @tyler36 if you could go through again, @khalwat if you could make sure your concerns are taken care of.

@susannemoog I'd still love to have you take a look.  

Artifacts for testing are in https://github.com/drud/ddev/pull/4197#issuecomment-1241409569

File: docs/content/users/quickstart.md [hunk 4]
```diff
@@ -312,17 +312,16 @@ DDEV comes ready to work with any PHP project, and has deeper support for severa
...
 
-    The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) just as it can for Laravel.
+    The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) just as it can for Laravel. DDEV automatically updates or creates the .env file with the database information.
 
     === "Composer"
         ```bash
         mkdir my-laravel-app
         cd my-laravel-app
-        ddev config --project-type=laravel --docroot=public --create-docroot
+        ddev composer create --prefer-dist --no-install --no-scripts laravel/laravel
```
Comment (tyler36) [inline docs/content/users/quickstart.md:321]:
```suggestion
        ddev config --project-type=laravel --docroot=public --create-docroot
        ddev composer create --prefer-dist --no-install --no-scripts laravel/laravel
```

**Comment by rfay** [comment]
Happy to have feedback from reviewers, just use the HEAD version! Thanks for all your input!

**Comment by stasadev** [comment]
I found a bug, ddev is supposed to look for an exact match in the `.env`, but it looks for any line that contains that variable.

For example, I have `DYNAMODB_CONNECTION` in my `.env` (from [this](https://github.com/baopham/laravel-dynamodb/blob/963811691313124e7562a02aefdcc31fdbd75e78/config/dynamodb.php#L14) package) and ddev thinks it is the same as `DB_CONNECTION`, and replaced my config with `DYNAMODB_CONNECTION="mysql"`.

**Comment by rfay** [comment]
Outstanding, thanks @stasadev - I'll work on that today.

**Pull request merged**